### PR TITLE
[FSDP] Enable mixed hybrid/non-hybrid sharding strategies

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`
* #90871 [FSDP][4/N] Refactor func to share state/init handle attrs
* #90862 [FSDP][3/N] Move `fsdp_modules(root_only=True)` -> `_get_fsdp_root_states()`
* #90861 [FSDP][2/N] Move `fsdp_modules(root_only=False)` -> `_get_fsdp_states()`
* #90864 [FSDP][Easy] Rename `entry` -> `fsdp_module` to be more descriptive
* #90860 [FSDP][1/N] Add `_get_fsdp_states()`
* **#90846 [FSDP] Enable mixed hybrid/non-hybrid sharding strategies**
* #90859 [FSDP][Easy] Use `run_subtests` for hybrid shard test
* #90858 [FSDP][Easy] ufmt files
* #90840 [FSDP][BE] Remove `_module_to_handles`, `HandleConfig`; use term "fqn"; clarify docs

In the context of hybrid sharding strategies, we only need to enforce the same process groups among the instances using a hybrid sharding strategy, not all instances. We can even mix and match the two different hybrid sharding strategies. This PR relaxes the validation to support this.

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 1]
```diff
@@ -1,19 +1,17 @@
 # Owner(s): ["oncall: distributed"]
 
 import contextlib
-from functools import partial
-from collections import Counter
 import sys
+from collections import Counter
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:5]: Could we have formatting changes separated out and in a different PR? Formatting changes with logical changes make it harder for reviewers to understand the critical parts of the PR to review.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 2]
```diff
@@ -54,6 +52,7 @@ def patch_allreduce(new_allreduce):
     finally:
         dist.all_reduce = orig_ar
 
+
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:55]: Admittedly nitpicky, but these sort of empty line changes also make the logical changes to the PR to review in my experience.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 3]
```diff
@@ -97,67 +97,58 @@ def test_raises_manual_wrap_hybrid_shard_when_none_policy(self):
...
-            model.lin1, process_group=(intra_pg, inter_pg), sharding_strategy=ShardingStrategy.HYBRID_SHARD
+            model.lin1,
+            process_group=(intra_pg, inter_pg),
+            sharding_strategy=ShardingStrategy.HYBRID_SHARD,
         )
         model = FSDP(
-            model, process_group=(intra_pg, dist.new_group()), sharding_strategy=ShardingStrategy.HYBRID_SHARD
+            model,
+            process_group=(intra_pg, dist.new_group()),
+            sharding_strategy=ShardingStrategy.HYBRID_SHARD,
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:133]: nit: these formatting changes are another example where reviewers may need to read carefully to check if there are any logical changes to the test.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 4]
```diff
@@ -171,68 +162,137 @@ def test_fsdp_hybrid_shard_basic_setup(self):
...
-                    inp = fsdp_model.get_input(device=torch.cuda.current_device())
-                    out = fsdp_model(inp[0], inp[1])
-                    loss = fsdp_model.get_loss(inp, out)
-                    loss.backward()
-
-                num_flat_params = len(list(FSDP._fsdp_handles(fsdp_model)))
-                self.assertEqual(num_flat_params, cntr[orig_ar])
-                self.assertEqual(num_flat_params, cntr[orig_rs])
-                dist.barrier()
+        self.run_subtests(
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:165]: Ideally refactors such as switching to use `run_subtests` would be moved into a separate PR as they're unrelated to the description given in the PR.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 5]
```diff
@@ -171,68 +162,137 @@ def test_fsdp_hybrid_shard_basic_setup(self):
...
+            {
+                "hsdp_sharding_strategy": [
+                    ShardingStrategy.HYBRID_SHARD,
+                    ShardingStrategy._HYBRID_SHARD_ZERO2,
+                ],
+                "init_mode": ["uniform", "mixed"],
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:171]: Would it be better to have an enum / constant here instead of hardcoded strings?

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py [hunk 6]
```diff
@@ -171,68 +162,137 @@ def test_fsdp_hybrid_shard_basic_setup(self):
...
+                "device_id": torch.cuda.current_device(),
+                "sharding_strategy": hsdp_sharding_strategy,
+            }
+            fsdp_model = TransformerWithSharedParams.init(
+                self.process_group,
+                FSDPInitMode.RECURSIVE,
+                CUDAInitMode.CUDA_BEFORE,
+                fsdp_kwargs,
+            )
+        elif init_mode == "mixed":
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_hybrid_shard.py:271]: It looks like the difference between "uniform" and "mixed" is whether we mix and match HSDP sharding strategies. Could we better name these to indicate so?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 7]
```diff
@@ -37,37 +37,41 @@
...
+    Returns:
+        Tuple[HookState]: One of the instances' inter-node state (does not
+        matter which since they will share the same one).
+    """
+    intra_node_pgs = set()
+    inter_node_pgs = set()
+    inter_node_states = set()
+    for fsdp_module in root_module.fsdp_modules(root_module):
+        # TODO: Change this to handle's sharding strategy if we deprecate
+        # `ShardingStrategy` internally.
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_runtime_utils.py:56]: Do we have an issue tracking the discussions / decisions needed in order to deprecate `ShardingStrategy`?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 8]
```diff
@@ -108,16 +112,17 @@ def _lazy_init(
...
         if fsdp_module is root_module:
             continue
 
         if fsdp_module.sharding_strategy in HYBRID_SHARDING_STRATEGIES:
-            _validate_hybrid_shard_setup(state, fsdp_module)
             # Share the allreduce state across FSDP units. This is not strictly necessary
             # as each one already uses the same process group, but can slightly save memory
             # since other FSDP units allreduce state can be garbage collected.
-            fsdp_module._inter_node_state = state._inter_node_state
+            assert inter_node_state is not None, "Implementation error"
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_runtime_utils.py:124]: nit: can we have something more descriptive than implementation error, possibly hinting towards what the underlying issue may be?

File: torch/distributed/fsdp/_runtime_utils.py [hunk 9]
```diff
@@ -37,37 +37,41 @@
...
-            )
+    This checks that all instances using a hybrid sharding strategy have the
+    same intra- and inter-node process groups.
 
-        if fsdp_root._inter_node_pg != fsdp_module._inter_node_pg:
-            raise ValueError(
-                f"For {fsdp_root.sharding_strategy}, inter-node process groups do not match"
-            )
+    Returns:
+        Tuple[HookState]: One of the instances' inter-node state (does not
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_runtime_utils.py:48]: curious, I don't see `HookState` anywhere?

**Review by rohan-varma — COMMENTED** [review]
Apologies for the preemptive stamp - coming back for a full review :)

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 1 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py:5]: Sorry about that. I separated it out. Eventually, we should try to have all developers use ufmt so that we can ufmt files right before pushing. Converging to the same formatter saves the uncertainty of how code should be formatted, and we chose ufmt since PyTorch recommends that.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 3 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py]: Makes sense.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 4 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py:171]: Makes sense. I separated it into a different PR.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 5 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py]: Yes, good point. I refactored to an enum.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 6 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py]: I provided better names: `ALL_HYBRID_SHARD` and `MIXED_HYBRID_FULL_SHARD`.

File: torch/distributed/fsdp/_runtime_utils.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py]: Good catch. I recalled it being called that, but it is actually called `DefaultState`. We may want to revisit that name in the future.

File: torch/distributed/fsdp/_runtime_utils.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:58]: https://github.com/pytorch/pytorch/issues/90857

File: torch/distributed/fsdp/_runtime_utils.py (hunk 8 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py]: Added a detailed explanation.

File: test/distributed/fsdp/test_fsdp_hybrid_shard.py (hunk 1 above)
Comment (rohan-varma) [reply test/distributed/fsdp/test_fsdp_hybrid_shard.py:5]: Curious about how we will encourage this practice. `lintrunner` is today's automated tool that can run pre-commit and is enforced by PyTorch CI. 

Should we work with dev infra / CI folks to add ufmt to `lintrunner`? Without automation, enforcement of a linting standard is prone to breaking.

**Comment by awgu** [comment]
I will fix test failures tomorrow.
